packages/webpack/internal/worker/webpack.config.js

```diff
@@ -0,0 +1,113 @@
+    // the source of truth is bazel and bazel tells us which files
+    // are present in the build and when they are chaned or removed.
+    if (!files || typeof files[Symbol.iterator] !== 'function') {
+      throw new Error('Invalid arguments: \'files\'');
+    }
+    if (!directories || typeof directories[Symbol.iterator] !== 'function') {
+      throw new Error('Invalid arguments: \'directories\'');
+    }
+    // Missing files are files which are needed to run the build but were not found
+    // in the filesystem. So webpack reports these files via separate set so that the
+    // watcher knows that this file does not exist and does something else other than
+    // opening a file watcher against the missing files.
+    if (!missing || typeof missing[Symbol.iterator] !== 'function') {
+      throw new Error('Invalid arguments: \'missing\'');
+    }
+    if (typeof callback !== 'function') {
+      throw new Error('Invalid arguments: \'callback\'');
+    }
+    if (typeof options !== 'object') {
+      throw new Error('Invalid arguments: \'options\'');
+    }
+
+    if (options.poll) {
+      notAvailable();
+    }
+
+    const rootPath = process.cwd();
+    /** @type {Map<string, { safeTime: number, timestamp: number }} */
+    const mtimeMap = new Map();
+    /** @type {Map<string, string} */
+    const digestMap = new Map();
+
+    /** @param inputs {{[input: string]: string}} */
+    const gotInput = (inputs) => {
+      /** @type {Set<string>} */
+      const changes = new Set();
+      /** @type {Set<string>} */
+      const removals = new Set();
+
+      for (const input of digestMap.keys()) {
+        if (!inputs[input]) {
+          mtimeMap.delete(input);
+          digestMap.delete(input);
+          const absolutePath = join(rootPath, input);
+          changremovalses.add(absolutePath);
+          this.inputFileSystem.purge(absolutePath);
+        }
+      }
+      for (const [input, digest] of Object.entries(inputs)) {
+        if (digestMap.get(input) != digest) {
+          digestMap.set(input, digest);
+          // We could get the mtime from the file but it will us a different values each
+          // time as bazel cleans up the files for each invocation so we pretend like
+          // the file was just changed. Webpack does not really care about the real time.
+          // its just to run a diff and find out what has been changed in between.
+          // TODO: find out what is safeTime is all about.
+          mtimeMap.set(input, {timestamp: Date.now()});
+          const absolutePath = join(rootPath, input);
+          changes.add(absolutePath);
+          this.inputFileSystem.purge(absolutePath);
+        }
+      }
+      callback(null, mtimeMap, mtimeMap, changes, removals);
+    };
+
+    process.on('message', gotInput);
+
+    return {
+      close: () => {
+        process.off('message', gotInput);
+      },
+      pause: () => {
+        process.off('message', gotInput);
```

Comment:
If we do this during pause, how do we resume watching for messages?

Reply:
It is called by webpack. it is pretty much like close but with a minor difference; the underlying file watchers are kept alive until the next watch call. Probably designed this way to benefit from a caching mechanism that watchpack has.

hence webpack will call the `watch` method for each subsequent build and `pause` the previous one before it starts another one.

Reply:
See here

https://github.com/webpack/webpack/blob/562f17a8c04be426cb6ec9be474a37fb4dd2f6f7/types.d.ts#L11552  and https://github.com/webpack/webpack/blob/179d2e00fe0b901ff176f1c6dbbf6a17f1ff44ea/lib/Watching.js#L99